File: 2D_Matrix_search.java
// Time Complexity : M + log(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
import java.util.*;

class matrix_Search {

    public  static boolean searchMatrix(int[][] matrix, int target) 
        {
            int row = matrix.length;
            int column = matrix[0].length;

            int start = 0;
            int end = row-1;
            int mid = 0;
            int targetRow = 0;

            while(start <= end) 
            {
                if(matrix[start][column-1] >= target) 
                {
                    targetRow = start;
                    break;
                }
                else 
                {
                    start++;
                }
            }

            start = 0;
            end = column-1;

            while(start <= end) 
            {
                mid = start + (end-start)/2;
                if(matrix[targetRow][mid] == target)
                    return true;
                else if(matrix[targetRow][mid] < target)
                {
                    start = mid+1;
                }
                else
                {
                    end = mid-1;
                }
            }
            return false;
        }
   
    public static void main(String args[]) 
        {
           int arr[][] = {{1,3,5,7}, {10,11,16,20}, {23,30,34,60}};
           boolean result;
           result = searchMatrix(arr, 1);
           if (result == true) {
               System.out.println("Target found");
           }
           else {
               System.out.println("Target not found");
           }
        }
}

File: 2DMatrixSearch.cpp
// Time Complexity : O(logm*n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int start = 0; 
        int end = matrix.size()-1; 

        while(start <= end) {
            int mid = (start+end)/2;
            if(target >= matrix[mid][0]) {
                int midEnd = matrix[mid].size()-1; 
                if(target <= matrix[mid][midEnd]) {
                    return vanillaBinarySearch(matrix[mid], target);
                } else start = mid+1; 
            } else end = mid-1; 
        }

        return false;     
    }

    bool vanillaBinarySearch(vector<int>& nums, int target) {
        int start = 0; 
        int end = nums.size()-1; 

        while(start <= end) {
            int mid = (start+end)/2; 
            if(target == nums[mid]) return true; 
            if(target < nums[mid]) end = mid-1; 
            else start = mid+1; 
        }

        return false; 
    }
};